### PR TITLE
[To rel/0.12] [IOTDB-2532] Query with align by device can't get value after clear cache

### PR DESCRIPTION
**Pull request by Alima777**
Path in where clause is not included in measurements of specific device.

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/qp/physical/crud/RawDataQueryPlan.java [hunk 1]
```diff
@@ -54,6 +56,19 @@ public IExpression getExpression() {
 
   public void setExpression(IExpression expression) throws QueryProcessException {
     this.expression = expression;
+    updateDeviceMeasurementsUsingExpression(expression);
+  }
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/qp/physical/crud/RawDataQueryPlan.java:60]: * `setExpression` will be called twice, once in `transformFilterOperatorToExpression` of `QueryOperator` which generating logical plan, once in `QueryRouter` after optimizing this expression. Although it won't make it it doesn't affect its correctness, the performance may be worse than before.
* In align by device query, the `deviceToMeasurements` should only contains current device, in your current implementation, all the devices will be saved in `deviceToMeasurements`, even if the devices that has been traversed.

**Review by JackieTien97 — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/apache/iotdb/db/qp/physical/crud/RawDataQueryPlan.java (hunk 1 above)
Comment (Alima777) [reply server/src/main/java/org/apache/iotdb/db/qp/physical/crud/RawDataQueryPlan.java:60]: Oh, Why don't we move `optimizing this expression` from `QueryRouter` to `QueryOperator` when generating this expression, then setExpression will be invoked only once.

Comment (Alima777) [reply server/src/main/java/org/apache/iotdb/db/qp/physical/crud/RawDataQueryPlan.java:60]: > `all the devices will be saved in deviceToMeasurements`
That's not right. In align by device query, there is a map `<device -> expression>`, so one expression will only include current device.

Comment (JackieTien97) [reply server/src/main/java/org/apache/iotdb/db/qp/physical/crud/RawDataQueryPlan.java:60]: > Oh, Why don't we move `optimizing this expression` from `QueryRouter` to `QueryOperator` when generating this expression, then setExpression will be invoked only once.

It sounds ok, you can move the optimization logic from `QueryRouter` to `transformQuery` function of  `PhysicalGenerator`. I found that only `rawQuery`, `aggregate`, `groupBy` and `udtfQuery` call the `optimize` function, you can figure it out whether other queries can apply that `optimize`.

Comment (JackieTien97) [reply server/src/main/java/org/apache/iotdb/db/qp/physical/crud/RawDataQueryPlan.java:60]: > > `all the devices will be saved in deviceToMeasurements`
> 
> That's not right. In align by device query, there is a map `<device -> expression>`, so one expression will only include current device.

I see, although different devices share the same `RawDataQueryPlan`, `setDeduplicatedPaths` of `RawDataQueryPlan` will call `clear()` function of `deviceToMeasurements`.

**Review by JackieTien97 — CHANGES_REQUESTED** [review]

**Pull request merged**